lightning/src/ln/channelmanager.rs

```diff
@@ -7202,6 +7207,133 @@ where
+							payment_paths, payment_hash, created_at
+						);
+						match builder.and_then(|b| b.allow_mpp().build_and_sign(secp_ctx)) {
+							Ok(invoice) => Some(OffersMessage::Invoice(invoice)),
+							Err(error) => Some(OffersMessage::InvoiceError(error.into())),
+						}
+					},
+					Ok((payment_hash, _payment_secret)) => {
+						// TODO: Include payment_secret in payment_paths.
+						let payment_paths = vec![];
+						#[cfg(not(feature = "no-std"))]
+						let builder = invoice_request.respond_with(payment_paths, payment_hash);
+						#[cfg(feature = "no-std")]
+						let created_at = Duration::from_secs(
+							self.highest_seen_timestamp.load(Ordering::Acquire) as u64
+						);
+						#[cfg(feature = "no-std")]
+						let builder = invoice_request.respond_with_no_std(
+							payment_paths, payment_hash, created_at
+						);
+						let response = builder.and_then(|builder| builder.allow_mpp().build())
+							.map_err(|e| OffersMessage::InvoiceError(e.into()))
+							.and_then(|invoice|
+								match invoice.sign(|digest, tag, bytes, metadata|
+									self.node_signer.sign_bolt12_message(digest, tag, bytes, metadata)
+								) {
+									Ok(invoice) => Ok(OffersMessage::Invoice(invoice)),
+									Err(SignError::Signing(())) => Err(OffersMessage::InvoiceError(
+											InvoiceError::from_str("Failed signing invoice")
+									)),
+									Err(SignError::Verification(_)) => Err(OffersMessage::InvoiceError(
+											InvoiceError::from_str("Failed invoice signature verification")
+									)),
+								});
+						match response {
+							Ok(invoice) => Some(invoice),
+							Err(error) => Some(error),
+						}
+					},
+					Err(()) => {
+						Some(OffersMessage::InvoiceError(SemanticError::InvalidAmount.into()))
+					},
+				}
+			},
+			OffersMessage::Invoice(invoice) => {
+				if !invoice.verify(expanded_key, secp_ctx) {
+					Some(OffersMessage::InvoiceError(InvoiceError::from_str("Unrecognized invoice")))
+				} else if invoice.features().requires_unknown_bits() {
+					Some(OffersMessage::InvoiceError(SemanticError::UnknownRequiredFeatures.into()))
+				} else {
+					let recipient_onion = RecipientOnionFields {
+						payment_secret: None,
+						payment_metadata: None,
+					};
+					let payment_hash = invoice.payment_hash();
+					let payment_id = PaymentId(payment_hash.0);
+					let route_params = RouteParameters {
+						payment_params: PaymentParameters::from_bolt12_invoice(&invoice),
+						final_value_msat: invoice.amount_msats(),
+					};
+					if let Err(e) = self.send_payment(
+						payment_hash, recipient_onion, payment_id, route_params,
+						BOLT_12_INVOICE_RETRY_STRATEGY
+					) {
+						// TODO: Should we send an invoice_error?
+						log_error!(self.logger, "Failed paying invoice: {:?}", e);
+					}
+					None
+				}
+			},
+			OffersMessage::InvoiceError(invoice_error) => {
+				log_error!(self.logger, "Received invoice_error: {}", invoice_error);
```

Comment:
Can we correlate it with an outbound payment? It might not even be for one (i.e., if it is in response to an invoice message).

Reply:
Ah, I neglected to re-read the full context here. So, I think this requires having the `path_id` in the onion message's reply path (for the `InvoiceRequest`) contain the payment id. We'd also need to pipe this through to `OffersMessageHandler::handle_message`.

Not sure if you prefer to go down this path at the moment. I have it using `log_trace` for now. In practice, we'd would generate an `InvoiceRequestFailed` immediately instead of waiting the constant three timer ticks for the request to timeout.